# Step 1 - Data consistency steps

In [1]:
#Importing libraries
import pandas as pd
import numpy
import os

In [2]:
#Path variable
path = r'C:\Users\Arjun\Documents\Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\Arjun\\Documents\\Instacart Basket Analysis'

In [4]:
# Products data frame
df_prods = pd.read_csv(os.path.join(path, 'Data','Original Data','products.csv'), index_col = False)

In [5]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [6]:
#Wrangled Orders data frame
df_ords = pd.read_csv(os.path.join(path,'Data','Prepared Data','orders_wrangled.csv'), index_col = False)

In [7]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0           0   2539329        1             1                  2   
1           1   2398795        1             2                  3   
2           2    473747        1             3                  3   
3           3   2254736        1             4                  4   
4           4    431534        1             5                  4   

   order_hour_of_day  days_since_last_order  
0                  8                    NaN  
1                  7                   15.0  
2                 12                   21.0  
3                  7                   29.0  
4                 15                   28.0

In [8]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_last_order  
count       3.421083e+06       3.421083e+06           3.214874e+06  
mean        2.776219e+00       1.345202e+01           1.111484e+01  
std         2.046829e+00       4.226088e+00           9.206737e+00  
min         0.000000e+00       0.000000e+00           0.000000e+00  
25%         1.000000e+00       1.000000e+01           4.000000e+00  
50%         3.000000e+00       1.300000e+01           7.000000e+00  
75%         5.000000e+00       1.600000e+01           1.500000e+01  
max         6.000000e+00       2.300000e+01           3.000000e+01

## Mixed Type Data

In [9]:
df_test = pd.DataFrame()

In [10]:
df_test['mix'] = ['a','b',1,True]

In [11]:
#mixed type column
df_test.head()

mix
0     a
1     b
2     1
3  True

In [12]:
#Logic for checking mixed types in data frames
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix



## Missing Values

In [13]:
#Finding the missing values in the df_prods data frame
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
#Since there are 16 missing values in the product name, we will do subsetting on those observations
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [ ]:
#Addressing Missing Values
# 1) Using the flag
# 2) Imputing the missing values using mean or median
df['column with missings'].fillna(mean value, inplace=True)
 OR 
df['column with missings'].fillna(median value, inplace=True)

In [17]:
df_prods.shape

(49693, 5)

In [18]:
#Creating a clean data frame which does not contain missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods_clean.shape

(49677, 5)

In [ ]:
#Another logic to remove all missing values is using drop() method
df_prods.dropna(subset = ['product_name'], inplace = True)

## Duplicates

In [20]:
#Finding the duplicate rows of the df_prods_clean dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
#Addressing the duplicate rows
df_prods_clean.shape

(49677, 5)

In [23]:
#Using the dopr duplicates method
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

## Exporting the Cleaned dataset

In [ ]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data','Prepared Data','products_checked.csv'))

## Step 2

In [25]:
#Descriptive Statistics
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_last_order  
count       3.421083e+06       3.421083e+06           3.214874e+06  
mean        2.776219e+00       1.345202e+01           1.111484e+01  
std         2.046829e+00       4.226088e+00           9.206737e+00  
min         0.000000e+00       0.000000e+00           0.000000e+00  
25%         1.000000e+00       1.000000e+01           4.000000e+00  
50%         3.000000e+00       1.300000e+01           7.000000e+00  
75%         5.000000e+00       1.600000e+01           1.500000e+01  
max         6.000000e+00       2.300000e+01           3.000000e+01

The descriptive statistics for orders dataset looks to be fine on the surface. We will need to do more in-depth analysis to understand the consequences.

In [26]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The products dataset has a max price value of 99999. This is very high as compared to the mean value. This might need a check as to which product has such a higher price.

In [27]:
#We can check that row of max price in the products dataset
df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

## Step 3 & 4 - Mixed Data Types

In [31]:
#df_ords mixed data type check
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)
    
print('No Mixed types within the data set')

No Mixed types within the data set


In [34]:
df_ords.dtypes

Unnamed: 0                 int64
order_id                   int64
user_id                    int64
order_number               int64
order_day_of_week          int64
order_hour_of_day          int64
days_since_last_order    float64
dtype: object

**There are no mixed data types within the Orders data set**

## Step 5 - Missing Values

In [35]:
# Checking for missing values in Orders dataset
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
order_day_of_week             0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [36]:
#Removing the unnamed column in df_ords dataset
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [38]:
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

In [39]:
df_Nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [40]:
df_Nan

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
11        2168274        2             1                  2   
26        1374495        3             1                  1   
39        3343014        4             1                  6   
45        2717275        5             1                  3   
...           ...      ...           ...                ...   
3420930    969311   206205             1                  4   
3420934   3189322   206206             1                  3   
3421002   2166133   206207             1                  6   
3421019   2227043   206208             1                  1   
3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 6 columns]

In [41]:
df_Nan.shape

(206209, 6)

From the above steps, we can see that the days_since_last order columns have missing values in the Orders dataframe. Upon further querying the particular columns, we see that there are 206209 missing values.

## Step 6

As we can see from the above queries, that there are 206209 missing values in the Subset of df_ords dataset. Also, there are user_ids with the same numbers in the total dataset of df_Nan. This signifies that the days_since_last_order will be **NaN for the first order** which makes correct sense. This does give an understanding of the reasoning behind having missing values in the dataset which makes sense.

## Step 7 - Duplicate Value checks

In [44]:
#duplicate value for df_ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [45]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

In [46]:
df_ords_dups.shape

(0, 6)

In [47]:
df_ords_dups.size

0

## Step 8

There are **No duplicates found** in the df_ords dataset.

## Step 9 - Export

In [48]:
#Orders dataset export
df_ords.to_csv(os.path.join(path,'Data','Prepared Data','orders_checked.csv'))

In [49]:
#Products dataset export - I am using the cleaned file
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data', 'Prepared Data','produtcs_checked.csv'))